Trivial change arf

ezyang opened the pull request:
Pull Request resolved: https://github.com/pytorch/pytorch/pull/{}

none

ezyang commented on torch/csrc/autograd/jit_closure.cpp:
```diff
@@ -0,0 +1,244 @@
+};
+
+// Used for inputs to the closure
+struct Placeholder : public Function {
```
I think Replicate strictly generalizes Placeholder.

apaszke replied:
While what you're saying is true (it's even equivalent to Identity that we already have!), I did it just so that the node type tells something about it's purpose. I'm ok with having these additional 5 lines just so that we have a specialized node type that can be used for this particular purpose

ezyang replied:
Sure. One thing that is a little irritating is if a parameter is used multiple times, you'll have Placeholder, and then a Replicate immediately after. But that shouldn't be breaking the perf bank, so 👍

apaszke replied:
Yeah, we generally optimize out replicates except for the inputs, but it should have no perf impact and makes life much easier

ezyang commented on torch/csrc/autograd/jit_closure.cpp:
```diff
@@ -0,0 +1,246 @@
+    }
+  }
+
+  // TODO: we could probably call into some of our C functions in here to make it faster
```
IIRC, you're referring to `THPFunction_apply` (because a Python Function's apply is always implemented using it)?

apaszke replied:
It's implemented like this in most cases. We don't forbid overloading it, which is e.g. what backward for Python Functions does. However these backward nodes should not get there so it's not a problem. But it might break in the future

ezyang commented on torch/csrc/autograd/jit_closure.cpp:
```diff
@@ -0,0 +1,244 @@
+    scalar_args.reserve(op->scalar_args.size());
+    for (auto& arg_ptr : op->scalar_args) {
+      Py_INCREF(arg_ptr.get());
+      scalar_args.emplace_back(arg_ptr.get());
```
Some day we'll add a helper method to do this :)

ezyang commented on torch/csrc/autograd/jit_closure.cpp:
```diff
@@ -0,0 +1,244 @@
+    return py_inputs;
+  }
+
+  variable_list unpackOutputs(THPObjectPtr& result) {
```
Ugh, it's too bad we can't const this (I assume the problem is GET_SIZE/GET_ITEM aren't const, right?)

apaszke replied:
Maybe? I haven't even thought about that. GET_SIZE/GET_ITEM might be ok, because these are macros IIRC.

ezyang commented on torch/csrc/autograd/jit_closure.cpp:
```diff
@@ -0,0 +1,244 @@
+
+std::unique_ptr<AutogradClosure> createAutogradClosure(Graph *graph) {
+  std::unordered_map<Node*, std::shared_ptr<Function>> node_map;
+  std::unique_ptr<AutogradClosure> result {new AutogradClosure()};
```
I don't think it's good to return a unique_ptr that doesn't know how to dispose itself. There's only one call-site for createAutogradClosure so it seems better to just return the wrapped thing right off the bat.

apaszke replied:
What do you mean? It knows what to do (uses `std::default_delete`, which calls `delete`)

ezyang replied:
You are right! Can't we use `std::make_unique` then?

apaszke replied:
`std::make_unique` is a C++14 thing :blush:

ezyang commented on torch/csrc/autograd/python_function.h:
```diff
@@ -31,6 +32,20 @@ struct PyFunction : public Function {
+  PyTuple_SET_ITEM(tuple, 0, obj.release());
+  obj = tuple;
+  return true;
+}
```
This refactor is at high risk of conflicting with changes to master.

apaszke replied:
Not really. No one should be touching this part (why would you modify this function?) and if they modify the header, it will be a simple fix during rebase

ezyang commented on torch/csrc/autograd/jit_closure.h:
```diff
@@ -0,0 +1,18 @@
+
+namespace torch { namespace autograd {
+
+struct AutogradClosure {
```
I'm not sure "closure" is the best name for this struct. In PL "closure" usually refers to a pair of both function and data. But there is no data in this case.

apaszke replied:
There's some data, because the functions (instructions/statements) are already parametrized. Closures still have inputs, which is what you called data in your comment. Does that make sense?

ezyang commented on torch/csrc/autograd/jit_closure.cpp:
```diff
@@ -0,0 +1,246 @@
+
+    IR_IF(node, PythonOp)
+      auto name = value->name();
+      // TODO: specialized ops will be probably handled by the tracer
```
I agree, this is probably not the right place to have this special case. Also, since you haven't implement MulBackward yet, this will break if you try to start tracing the backwards here.

apaszke replied:
Yea, but there's nothing I can do about that now ¯\_(ツ)_/¯

ezyang commented on torch/csrc/autograd/jit_closure.cpp:
```diff
@@ -0,0 +1,246 @@
+  std::vector<THPObjectPtr> scalar_args;
+};
+
+bool isMultireturn(Node * node) {
```
Let's put this in the IR proper.

ezyang commented on test/test_jit.py:
```diff
@@ -33,5 +34,20 @@ def test_lstm(self):
+
+        torch._C._tracer_enter((x, y))
+
+        z = torch.sigmoid(torch.tanh(x * (x + y)))
```
This test doesn't exercise the multi-return case.

ezyang commented:
I'm happy to merge this immediately. @zdevito, do you want to look before this goes in?

apaszke commented:
Rebased and enabled the test (+ added another output). Remember to squash the changes when you merge!

zdevito commented on torch/csrc/autograd/jit_closure.cpp:
```diff
@@ -0,0 +1,244 @@
+
+  // Iterate over nodes and inputs.
+  // NOTE: First ++ will skip the Return node.
+  auto nodes_it = ++graph->nodes().rbegin();
```
I think this would be clearer if it iterates over `inputs()` first and then over nodes. You can put the body in a named lambda: `auto handleNode= [&](Node * n) { ... }`

apaszke replied:
It has to iterate over nodes before inputs - all uses must be in node_map, so it needs to be a reverse topological order.

ezyang committed:
Pull Request resolved: https://github.com/pytorch/pytorch/pull/17

Pull Request resolved: https://github.com/pytorch/pytorch/pull/{}

none

Differential Version: 56221835
